### PR TITLE
Dbconfig fixes.

### PR DESCRIPTION
**Pull request by alainjobart**
Three main concerns:

1. CreadentialsServer was not always used. In particular, schema changes and backups were not using it. Re-organized the code a bit so it's easier to see where they're used, and fixed it.

2. mysqlctl and mysqlctld only basically start and stop mysqld. So they only need the dba user credentials. Fixed the code everywhere for that to work.

3. the mysql password was passed as a command line parameter to mysql / mysqladmin / mysql_upgrade, which is not secure (visible via PS, and in the logs). Now we only pass it when necessary, and we put it in a temporary config file.

### CODE REVIEW

**Comment by alainjobart** [comment]
@pivanof ready for review, thanks. Will squash all after review and comments.

File: examples/local/vttablet-up.sh [hunk 1]
```diff
@@ -74,7 +75,7 @@ for uid_index in $uids; do
   fi
   $VTROOT/bin/mysqlctl \
     -log_dir $VTDATAROOT/tmp \
-    -tablet_uid $uid $dbconfig_flags \
+    -tablet_uid $uid $dbconfig_dba_flags \
```
Comment (pivanof) [inline examples/local/vttablet-up.sh]: Please put this on a new line since it's not related to -tablet_uid.

File: examples/local/vttablet-down.sh [hunk 2]
```diff
@@ -27,7 +27,6 @@ for uid_index in $uids; do
 
   echo "Stopping MySQL for tablet $alias..."
   $VTROOT/bin/mysqlctl \
-    -db-config-dba-uname vt_dba \
```
Comment (pivanof) [inline examples/local/vttablet-down.sh]: Why did you remove it from here?

File: go/vt/mysqlctl/mysqld.go [hunk 3]
```diff
@@ -71,53 +69,49 @@ type Mysqld struct {
...
-		dba:                dba,
-		allprivs:           allprivs,
-		dbApp:              app,
-		dbaPool:            dbaPool,
-		appPool:            appPool,
-		replParams:         repl,
-		dbaMysqlStats:      dbaMysqlStats,
-		allprivsMysqlStats: allprivsMysqlStats,
-		tabletDir:          path.Dir(config.DataDir),
+func NewMysqld(config *Mycnf, dbcfgs *dbconfigs.DBConfigs, enablePublishStats bool, flags dbconfigs.DBConfigFlag) *Mysqld {
```
Comment (pivanof) [inline go/vt/mysqlctl/mysqld.go]: Maybe put flags after dbcfgs since they are related to each other? And also it should be dbconfigFlags to be more understandable, right?

File: go/vt/mysqlctl/mysqld.go [hunk 4]
```diff
@@ -71,53 +69,49 @@ type Mysqld struct {
 
 // NewMysqld creates a Mysqld object based on the provided configuration
 // and connection parameters.
-func NewMysqld(config *Mycnf, dba, allprivs, app, repl *sqldb.ConnParams, enablePublishStats bool) *Mysqld {
-	noParams := sqldb.ConnParams{}
-	if *dba == noParams {
-		dba.UnixSocket = config.SocketFile
```
Comment (pivanof) [inline go/vt/mysqlctl/mysqld.go]: Why this is not needed anymore?

File: go/vt/mysqlctl/mysqld.go [hunk 5]
```diff
@@ -157,18 +152,16 @@ func (mysqld *Mysqld) RunMysqlUpgrade() error {
...
-	// run the program, if it fails, we fail
+	// Run the program, if it fails, we fail.  Note in this
+	// moment, mysqld is running with no grant tables on the local
+	// socket only, so this doesn't need any password for root.
 	args := []string{
 		// --defaults-file=* must be the first arg.
 		"--defaults-file=" + mysqld.config.path,
 		"--socket", mysqld.config.SocketFile,
-		"--user", mysqld.dba.Uname,
+		"--user", "root",
```
Comment (pivanof) [inline go/vt/mysqlctl/mysqld.go]: So you decided to not use "dummy_user"? Why? Could be a nice way to document that authentication doesn't really work at this point.

File: go/vt/mysqlctl/mysqld.go [hunk 6]
```diff
@@ -688,7 +687,12 @@ func deleteTopDir(dir string) (removalErr error) {
 // executeMysqlCommands executes some SQL commands,
 // using the mysql command line tool.
 func (mysqld *Mysqld) executeMysqlCommands(connParams *sqldb.ConnParams, sql string) error {
-	return mysqld.executeMysqlScript(connParams, strings.NewReader(sql))
+	params, err := dbconfigs.WithCredentials(connParams)
```
Comment (pivanof) [inline go/vt/mysqlctl/mysqld.go]: Why this call to WithCredentials() is not inside executeMysqlScript()?

File: test/tablet.py [hunk 7]
```diff
@@ -613,14 +619,15 @@ def wait_for_mysqlctl_socket(self, timeout=60.0):
...
+  def _add_dbconfigs(self, cfg, args, repl_extra_flags=None):
     if repl_extra_flags is None:
       repl_extra_flags = {}
-    config = dict(self.default_db_config)
-    if self.keyspace:
-      config['app']['dbname'] = self.dbname
-      config['repl']['dbname'] = self.dbname
-    config['repl'].update(repl_extra_flags)
+    config = dict(cfg)
+    if 'repl' in config:
```
Comment (pivanof) [inline test/tablet.py]: Why did you add "if" here?

File: go/vt/mysqlctl/mysqld.go (hunk 6 above)
Comment (alainjobart) [inline go/vt/mysqlctl/mysqld.go]: Because the other caller of executeMysqlScript doesn't want to add credentials. Which come to look at it may be a problem, let me see.

Comment (alainjobart) [inline go/vt/mysqlctl/mysqld.go]: The other use case is right after Init, we just want to connect as user root with no password, so we don't want to go through credentials server. I think that makes sense.

Comment (bbeaudreault) [inline go/vt/mysqlctl/mysqld.go]: I know the no-password root user is depended on elsewhere as well, so it's fair if we want to punt on that in this PR. I was planning on digging into this internally, and seeing if it made sense to PR it up ourselves or if we could add the no-password root user back in.

Comment (alainjobart) [inline go/vt/mysqlctl/mysqld.go]: I think that's independent, actually. We only use the root user in two situations:

1. we have just created an empty database from scratch, it has no users, and we're about to run the init sql script. Since there is no user configured, we have to use root without password.

2. we are restoring from backup, and we want to run mysql_upgrade. In that case, we run mysql with --skip-grant-tables and --skip-networking, so we don't need the password, but the database is not visible from the outside either.

In both these cases, your uses root password is not needed at all. Does that make sense?

We never use the root user otherwise, as a hardcoded 'root' user. Of course, you can configure your dba user name to be root (instead of vt_dba) and password to be set, and we'll use that when doing DBA tasks.

Comment (bbeaudreault) [inline go/vt/mysqlctl/mysqld.go]: Those two usages make sense. The place where I ran into this was when creating a Backup, it stops and then starts mysql after the backup is finished.  After starting, it calls mysqld.Wait, which uses the root user.  https://github.com/youtube/vitess/blob/master/go/vt/mysqlctl/mysqld.go#L288

File: go/vt/mysqlctl/mysqld.go (hunk 5 above)
Comment (pivanof) [inline go/vt/mysqlctl/mysqld.go]: It definitely doesn't have to exist. With --skip-grant-tables MySQL doesn't even know which users exist or not. So if something failed for you it wasn't related to the existence of the user.

File: go/vt/mysqlctl/mysqld.go (hunk 6 above)
Comment (pivanof) [inline go/vt/mysqlctl/mysqld.go]: Please add a comment to this function as well then to better stress the difference between executeMysqlCommands and executeMysqlScript. Probably the comment on executeMysqlScript should be expanded as well to point out that if adding credentials is needed then executeMysqlCommands should be used, though then it becomes awkward due to difference in types of input arguments. :(

File: test/tablet.py (hunk 7 above)
Comment (pivanof) [inline test/tablet.py]: Then this is a half-baked and hackish solution, because existence of config['app'] is not related to existence of config['repl'] in any way. It would be great to extract that into a separate "if".

File: go/vt/mysqlctl/mysqld.go (hunk 6 above)
Comment (alainjobart) [inline go/vt/mysqlctl/mysqld.go]: executeMysqlCommands is only used by the schema side, with dba credentials. So I moved it there, and renamed it. Much clearer now.

File: test/tablet.py (hunk 7 above)
Comment (alainjobart) [inline test/tablet.py]: OK. I now have 3 ifs instead of one.

File: go/vt/mysqlctl/mysqld.go (hunk 5 above)
Comment (alainjobart) [inline go/vt/mysqlctl/mysqld.go]: So with '--user dummy_user' it tells me mysql_upgrade fails to run, and it echoes a permission denied message. With vt_dba or root, it works. So if a user is specified, it has to exist.

However, if I don't specify a user, it works. So I'll just do that.

File: test/tablet.py (hunk 7 above)
Comment (pivanof) [inline test/tablet.py]: 3 "if"s, but the intention is clear and there's no hidden dependencies.
